roxctl/README.md

```diff
@@ -0,0 +1,189 @@
+
+## Creating an alias for deprecated values
+
+When changing flags or commands, aliases must be in place to avoid breaking users of said command.
+
+Currently, two options exist for this:
+- Cobra's [alias feature](https://pkg.go.dev/github.com/spf13/cobra#Command).
+- Handling the alias in-code.
+
+For Cobra's alias feature, the following limitations currently apply:
+- Renaming a command is straight-forward **as long as the command hierarchy does not change**.
+- Creating an alias for a flag currently is not supported, only for args.
+
+In case your change falls under the limitations of the cobra alias feature, it's recommended to handle the alias within
+the command's code itself.
+
+The below code highlights changes for a command:
+```go
+package sample
+
+func OldCommand(cliEnvironment environment.Environment) *cobra.Command {
+	sampleCmd := &sampleCmd{env: cliEnvironment}
+
+	cmd := &cobra.Command{
+		Use: "old-sample",
+		Short: "A sample command",
+		Long: `A sample command.
+This is a sample for a multi-line detailed description for the sample command`,
+        Hidden: true,
+		Args: cobra.NoArgs,
+		// Each command **must** always use `RunE` and return an error.
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cliEnvironment.Logger().WarnLn("old-sample is deprecated. Please use new-sample.")
+			return sampleCmd.RunE(cmd, args)
+		},
+	}
+
+	cmd.Flags().StringVarP(&sampleCmd.name, "name", "n", "", "name to say hello to")
+    return cmd
+}
+
+func NewCommand(cliEnvironment environment.Environment) *cobra.Command {
+	sampleCmd := &sampleCmd{env: cliEnvironment}
+
+	cmd := &cobra.Command{
+		Use: "new-sample",
+		Short: "A sample command",
+		Long: `A sample command.
+This is a sample for a multi-line detailed description for the sample command`,
+		Args: cobra.NoArgs,
+		// Each command **must** always use `RunE` and return an error.
+		RunE: sampleCmd.RunE,
+	}
+
+	cmd.Flags().StringVarP(&sampleCmd.name, "name", "n", "", "name to say hello to")
+	return cmd
+}
+
+// While the sample command now lives in the same package, it may live in a completely different package.
+type sampleCmd struct {
+	env environment.Environment
+
+	name string
+}
+
+// RunE will be used for the replaced command. It's a handy wrapper that is optional but preferred.
+func (s *sampleCmd) RunE(cmd *cobra.Command, args []string) error {
+	if err := sampleCmd.Construct(cmd); err != nil {
+		return err
+	}
+	return sampleCmd.sample()
+}
+
+// Construct ensure all flag values are resolved and possibly validated. Any additional external dependency that you may
+// require (e.g. a gRPC client) may be constructed here as well.
+func (s *sampleCmd) Construct(cmd *cobra.Command) error {
+	return nil
+}
+
+// Sample is an example for any possible business logic your command may execute.
+func (s *sampleCmd) sample() error {
+	s.env.Logger().InfofLn("Hello world, %s!", s.name)
+	return nil
+}
```

Comment:
I would ditch everything that is not related to renaming, so that readers don't need to figure what's boilerplate and what's relevant. You can keep function definitions to demonstrate that they are shared but maybe in some compact way like:
```
// Aliases can share code and reuse functions, e.g.:
func (s *sampleCmd) RunE(cmd *cobra.Command, args []string) error { ... }
func (s *sampleCmd) Construct(cmd *cobra.Command) error { ... }
func (s *sampleCmd) sample() error {. .. }
```